File: backend/app/schemas/areas.py
from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel, Field

from app.schemas.common import DataMode, SourceAttribution


class PostcodePoint(BaseModel):
    postcode: str
    latitude: float
    longitude: float
    country_code: str | None = None
    admin_district_code: str | None = None
    admin_ward_code: str | None = None
    source: SourceAttribution


class CrimeMonth(BaseModel):
    month: str
    count: int


class CrimeCategoryCount(BaseModel):
    category: str
    count: int


class CrimeSummary(BaseModel):
    status: DataMode
    latest_month: str | None = None
    latest_total: int | None = None
    months: list[CrimeMonth] = Field(default_factory=list)
    categories: list[CrimeCategoryCount] = Field(default_factory=list)
    source_url: str
    message: str | None = None


class PlanningConstraint(BaseModel):
    dataset: str
    name: str
    reference: str | None = None
    start_date: date | None = None


class PlanningSummary(BaseModel):
    status: DataMode
    total: int | None = None
    constraints: list[PlanningConstraint] = Field(default_factory=list)
    source_url: str
    message: str | None = None


class EPCCertificate(BaseModel):
    certificate_number: str
    address: str
    postcode: str | None = None
    current_rating: str | None = None
    lodgement_date: str | None = None
    uprn: str | None = None


class EPCSummary(BaseModel):
    status: DataMode
    total: int | None = None
    certificates: list[EPCCertificate] = Field(default_factory=list)
    source_url: str
    message: str | None = None


class FloodWarning(BaseModel):
    severity: str
    severity_level: int
    description: str
    area: str | None = None
    time_raised: datetime | None = None
    time_message_changed: datetime | None = None


class FloodSummary(BaseModel):
    status: DataMode
    total: int | None = None
    warnings: list[FloodWarning] = Field(default_factory=list)
    radius_km: int = 10
    source_url: str
    message: str | None = None


class AreaCheckResponse(BaseModel):
    postcode: PostcodePoint
    crime: CrimeSummary
    planning: PlanningSummary
    flood: FloodSummary
    limitations: list[str] = Field(default_factory=list)

File: frontend/types/areas.ts
import type { DataMode, SourceAttribution } from "./common";

export interface PostcodePoint {
  postcode: string;
  latitude: number;
  longitude: number;
  country_code: string | null;
  admin_district_code: string | null;
  admin_ward_code: string | null;
  source: SourceAttribution;
}

export interface CrimeSummary {
  status: DataMode;
  latest_month: string | null;
  latest_total: number | null;
  months: { month: string; count: number }[];
  categories: { category: string; count: number }[];
  source_url: string;
  message: string | null;
}

export interface PlanningSummary {
  status: DataMode;
  total: number | null;
  constraints: { dataset: string; name: string; reference: string | null; start_date: string | null }[];
  source_url: string;
  message: string | null;
}

export interface EPCCertificate {
  certificate_number: string;
  address: string;
  postcode: string | null;
  current_rating: string | null;
  lodgement_date: string | null;
  uprn: string | null;
}

export interface EPCSummary {
  status: DataMode;
  total: number | null;
  certificates: EPCCertificate[];
  source_url: string;
  message: string | null;
}

export interface FloodSummary {
  status: DataMode;
  total: number | null;
  warnings: {
    severity: string;
    severity_level: number;
    description: string;
    area: string | null;
    time_raised: string | null;
    time_message_changed: string | null;
  }[];
  radius_km: number;
  source_url: string;
  message: string | null;
}

export interface AreaCheckResponse {
  postcode: PostcodePoint;
  crime: CrimeSummary;
  planning: PlanningSummary;
  flood: FloodSummary;
  limitations: string[];
}
